Yoast/wordpress-seo
# P2 925 expand yoast head with json

## Context


* We want to output meta fields as a json object to assist 3rd party integrators.
* We want to prevent getting the same identical data twice, so a caching system needs to be implemented.

## Summary


This PR can be summarized in the following changelog entry:

* Adds yoast meta fields as a JSON object to the html head section

## Relevant technical choices:

* hook into `\register_rest_field` twice; once for the <meta> tags, once for the json output.
* this would make each presenter `get` the same data twice. To counter this performance hit, an `Abstract_Cached_Indexable_Presenter` decorator was introduced, which stores the output from the first time `get` is called in a protected field so the second time `get` is called the first result is reused.
* this creates a recursive loop in the decorator: `function get() { ( if x not set, call get-ohwait ) }` so all classes to be decorated `Abstract_Cached_Indexable_Presenter` are injected to the decorator itself, so the child->get can be called.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

* view post with ID 1 (or view any post and make note of its ID.
* view the source in the development console.
* notice that the document head contains (most of) these fields:

- canonical
- meta description
- article author
- article modified time
- article published time
- article publisher
- description
- image (if an image is present)
- locale
- site name
- title
- page type
- url
- rel next
- rel prev
- robots
- title
- twitter card
- twitter creator
- twitter description
- tiwtter image
- twitter site
- twitter title
- baidu
- bing
- google
- pinterest
- yandex

The presence of the above head fields indicates no regression bugs are introduced. 
Note that not always every field is present - this depends on page type / site configuration. To validate regression, you should run version 16.6 of the free plugin to see what fields are displayed for post #1 ( or your selected ID ); this PR should show the exact same fields with the exact same values for the same post viewed in v16.6.

and finally ( the reason for this PR ) 
visit http://[your-wordpress-server]/wp-json/wp/v2/posts/1 (or, instead of 1, enter the ID of the post you were using before.)
this will dump the content of the post to the screen.
you should copy and paste this entire document into a json prettifier e.g. https://jsonformatter.curiousconcept.com/ 
look for 2 properties: 
- yoast_head
- yoast_head_json 

- yoast_head should contain the same fields as found in the source in the previous test step. you should compare this with the output of release version 16.6 to make sure no regression bugs are added. Note that values may not always be identical because on the rest api some filters are not run and defaults behave differently. The main concern here is that this PR produces the same output as v16.6
- yoast_head_json should contain a JSON object consisting of the fields present in yoast_head; with identical values. 
The JSON keys should have been sanitized e.g. `og_twitter` instead of `og:twitter`, because : is an illegal character in json keys.

### Test instructions for QA when the code is in the RC


* [] QA should use the same steps as above.


QA can test this PR by following these steps:

* visit http://[your-wordpress-server]/wp-json/wp/v2/posts/
output starts with `[{"id":xxxxx` -> use xxxxx as the Id for the next steps.
* visit http://[your-wordpress-server]/wp-json/wp/v2/posts/ <Id>
* view the post on your wordpress site
* copy & paste the entire output in a json prettifier e.g. https://jsonformatter.curiousconcept.com/ 
* the json output should show a yoast_head property, containing tags with the the same meta content as the post.
* the json output should show a yoast_head_json property, with key / value versions of the 

This concerns the rest API of the wordpress site, and it could easily be verified in an automated test.

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

* rest api
* head <meta> tags

## UI changes


## Documentation


## Quality assurance


Fixes #15574
